### ISSUE TITLE
REMEMBERME only works once

### ISSUE THREAD

**Issue by fritzmg** [contao/contao#400]
**Affected version(s)**
4.7.1

**Description**  
The "Remember me" option at the login always only works for exactly _one_ auto login. When you auto logged in once, you will not be able to do so any more.

**How to reproduce**  

1. Go to https://demo.contao.org/en/login.html.
2. Klick on "Remember me" in the login form.
3. Log in as `j.smith` or `d.evans`.
4. Close the browser.
5. Go to https://demo.contao.org/en/login.html.
6. You will be logged in automatically.
7. Close the browser.
8. Go to https://demo.contao.org/en/login.html.

You will not be logged in. The `REMEMBERME` cookie will be deleted in the response header of **8.**:
```
set-cookie: REMEMBERME=deleted; expires=Thu, 15-Mar-2018 14:18:53 GMT; Max-Age=0; path=/; secure; httponly
```

**Comment by leofeyer** [comment]
Are you sure that the issue did not occur in Contao 4.6?

**Comment by fritzmg** [comment]
> Are you sure that the issue did not occur in Contao 4.6?

No, I only tested it in Contao 4.7.1 (since 4.6 is EOL).

**Comment by leofeyer** [comment]
Correct, although in this case it would have been interesting to know if we have introduced the defect in Contao 4.7, because then it might be cache or routing related. But we will figure it out anyway. 😄

**Comment by fritzmg** [comment]
I know, I just did not have the time to test it everywhere :)

If I get around to it, I'll report of course.

**Comment by xchs** [comment]
If it helps, I could test it in the afternoon with Contao 4.6.

**Comment by xchs** [comment]
I cannot reproduce it in Contao 4.6.

**Comment by leofeyer** [comment]
@Toflar: The issue does not occur in debug mode, therefore could be a header replay issue?

**Comment by Toflar** [comment]
Who knows...I haven't got the time to debug that atm.

**Comment by leofeyer** [comment]
This is the corresponding log entry:

`security.WARNING: The token storage was not populated with remember-me token as the RememberMeServices was not able to create a token from the remember me information. { "exception":"[object] (Symfony\\Component\\Security\\Core\\Exception\\CookieTheftException(code: 0): This token was already used. The account is possibly compromised.
`

So I guess it is a header replay issue. Will continue to debug…

**Comment by leofeyer** [comment]
Nope, it turns out that the ESI request to replace the `{{ua::class}}` insert tag causes the problem. This is due to the changes from #199, where we have added the missing `_scope` attribute to the controller reference in the `InsertTags` class (see #176).

**Comment by aschempp** [comment]
This could be because the new `forward` renderer that passes the cookies to the subrequest, which in turn then cause authentication to invalidate the token for security reasons. Can you check that path?

**Comment by leofeyer** [comment]
Correct. And the reason why it has worked before was the missing `_scope` attribute, which caused the `contao.routing.frontend_matcher` not to match the insert tag ESI request, so the firewall did not trigger the login routines.

Now that there is a `_scope` attribute, we need to find a way to either prevent rememberme logins in ESI requests or the ESI request has to use the updated cookie with the new serial, which is set by the main response.

**Comment by Toflar** [comment]
Okay, so not related to header-replay. The fact that it does work in `app_dev.php` is related to having no ESI at all :)

**Comment by leofeyer** [comment]
Correct.

**Comment by Toflar** [comment]
@aschempp why does the forward renderer handle ESI fragments anyway? It should only be able to handle regular fragments. You cannot expect other reverse proxies to handle ESI in such a way that they pass on request information. Or well the core one of Symfony should pass on the data according to the spec (https://www.w3.org/TR/esi-lang <-- scroll to "4 ESI Variables"). So you get cookies but you never get a request body or anything. You cannot handle POST values in an ESI fragment.
I feel like this is a bug that needs the up for discussion label and will remain unresolved for the time being.

**Comment by leofeyer** [comment]
I have added the "up for discussion" label.

> will remain unresolved for the time being

Since the next call is on April 11th, there will be a new Contao release (version 4.7.2) before we can discuss this. @contao/developers We should think about what is more important: a working rememberme implementation or a working maintenance mode implementation? Because we can only have one in Contao 4.7.2.

**Comment by Toflar** [comment]
Just leave it as is for now. Remember me is just broken.

**Comment by aschempp** [comment]
> why does the forward renderer handle ESI fragments anyway? It should only be able to handle regular fragments.

I think you are right, it does not. It "replaces" the `inline` renderer, not ESI.

> Now that there is a _scope attribute, we need to find a way to either prevent rememberme logins in ESI requests or the ESI request has to use the updated cookie with the new serial, which is set by the main response.

That's what we already built into header-replace

> what is more important: a working rememberme implementation or a working maintenance mode implementation? Because we can only have one in Contao 4.7.2.

How are these related? Did we add the scope attribute to support maintenance mode?

**Comment by leofeyer** [comment]
> Did we add the scope attribute to support maintenance mode?

Yes, we did: #176

**Comment by aschempp** [comment]
This, again, could easily be solved with the JWT 😞

**Comment by leofeyer** [comment]
As discussed in Mumble on April 11th, we want to open a Symfony issue and ask how they would solve this.

**Comment by leofeyer** [comment]
https://github.com/symfony/symfony/issues/31078

**Comment by leofeyer** [comment]
As discussed again in Mumble on May 9th, we want to try to fix this by reloading the page after a successful remember me authentication.

**Comment by Toflar** [comment]
Update: Reloading is not going to work either because fragments cannot reload ;-)
We need to adjust the series handling.
@fritzmg can you test if things work for you again with the latest header-replay-bundle? It might work now but in fact only by chance (because Symfony's ESI handling executes sub requests and thus the firewall is not initialized multiple times). So we still need to fix the series handling to make sure it also works for Varnish and Co.

**Comment by fritzmg** [comment]
~~Yeah, with terminal42/header-replay-bundle 1.5.3 this seems to work now :)~~

**Comment by fritzmg** [comment]
Sorry, got confused - no, the latest version did not fix the issue. Or did you mean dev-master?

**Comment by leofeyer** [comment]
This will eventually be fixed by #483.
